### PR TITLE
[TierTwo] Long Living Masternode Quorums - Part 2, network layer

### PR DESCRIPTION
**Pull request by furszy**
Built on top of #2607, #2632 and #2639. Focused on implementing the Masternode quorums network communication layer.

This layer is in charge of:
1) Open MN-to-MN, MN-to-Quorum, MN-to-MN_iqr (intra-quorum relay) and MN-to-MN-probe connections.
2) Implement a challenge-based connection authentication for MNs (new network message `mnauth` introduced).
3) Cache and store MN connections metadata such as the last successful connection time and the last connection attempt time.
4) Protect verified MN connections from eviction while they are part of the active quorum.
5) Send interest in plain LLMQ recovered sigs (`QSENDRECSIGS` net message).

Plus, added a new functional test `p2p_quorum_connect.py` which basically verifies the proper functioning of most of this points (except the eviction process).

And finally, bumped the protocol version to 70925 so peers can know to whom they can, or not, send the new p2p messages.

### CODE REVIEW

**Comment by furszy** [comment]
rebased on top of #2607, #2631,  #2632, #2639, #2649 and #2657. Plus worked further on the circular dependencies problems and removed most of the `masternode-sync.h` includes that were all around the sources.
Restructuring the masternode synchronization, decoupling the state into a separate class `tiertwo_sync_state` which is only in charge of storing and providing the tier two synchronization status, without any synchronization logic (the sync logic remains inside the masternode-sync class).

**Comment by furszy** [comment]
rebased on top of the recently rebased #2690. Once that one gets merged, this is ready to go.

File: src/rpc/client.cpp
Comment (random-zebra) [inline src/rpc/client.cpp:174]: `"llmq_type"`

File: src/rpc/client.cpp
Comment (random-zebra) [inline src/rpc/client.cpp:175]: this is a string too (`quorum_hash`). No need to convert it.

File: src/rpc/misc.cpp
Comment (random-zebra) [inline src/rpc/misc.cpp:784]: it looks like the first two arguments are required:
```suggestion
    if (request.fHelp || request.params.size() < 2 || request.params.size() > 4) {
```

File: src/rpc/misc.cpp
Comment (random-zebra) [inline src/rpc/misc.cpp:841]:
```suggestion
    { "hidden",             "mnconnect",              &mnconnect,              true,  {"conn_type", "mn_list", "llmq_type", "quorum_hash"} },
```

**Review by random-zebra — COMMENTED** [review]
Code ACK up to 3d9b2c13b4ae74616dd34413f11ec38b4d28ab59. Really nice job. Only nits and minor things so far.

**Comment by furszy** [comment]
Feedback tackled.

Plus:
-Improved the version msg transmitted data for regular connections.
-Added test coverage for the rejection of authenticated quorum members connections by regular peers.

File: test/functional/p2p_quorum_connect.py
Comment (random-zebra) [inline test/functional/p2p_quorum_connect.py:222]: this fixes the failure here https://github.com/PIVX-Project/PIVX/runs/4734568472

```suggestion
        with self.miner.assert_debug_log(["but we're not a masternode, disconnecting"]):
            no_version_node.send_message(msg_version(mn_challenge))
            time.sleep(2)
```

**Review by random-zebra — COMMENTED** [review]
Code ACK 0a33ab835cebdb4c9eddf18f78a324cad7b46131

File: src/net_processing.cpp
Comment (furszy) [reply src/net_processing.cpp:274]: yeah good, did it when the version msg was relaying the flag as well, then improved it removing the unneeded flag and forgot to remove it.

File: src/tiertwo/tiertwo_sync_state.h
Comment (furszy) [reply src/tiertwo/tiertwo_sync_state.h:27]: yeah better. Will clean it up.

File: test/functional/p2p_quorum_connect.py
Comment (furszy) [reply test/functional/p2p_quorum_connect.py:222]: that was me squashing last minute changes just before going to sleep hehe :v:

**Comment by furszy** [comment]
feedback tackled, plus rebased on master due conflicts with 2699.

**Review by random-zebra — APPROVED** [review]
ACK de3b51f772888ae377e95f73403d802031ba6d29

**Comment by furszy** [comment]
Dismissed the approval only to decrease the DMN connections thread sleep time up to the minimum to be able to support testnet/mainnet block spacing and dkg phases intervals.
Small code diff.

**Review by random-zebra — APPROVED** [review]
utACK fc707f0d09a1cad4b9fe5c79884a54de1502fc46 after force-push

File: test/functional/p2p_quorum_connect.py
Comment (Fuzzbawls) [inline test/functional/p2p_quorum_connect.py:57]: i had some pretty inconsistent results with hitting this timeout locally, an increase to `240` seemed to clear it up

**Review by Fuzzbawls — COMMENTED** [review]
looking pretty good, mostly minor nits and some styling or minor refactorings for feedback.

the 2 comments in `rpc/misc.cpp` can be addressed with this patch:
```diff
Index: src/rpc/misc.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/rpc/misc.cpp b/src/rpc/misc.cpp
--- a/src/rpc/misc.cpp	(revision 2e03a2c2bf3937b7c93c83f2cce7a6ab4fcd98c2)
+++ b/src/rpc/misc.cpp	(date 1642313165996)
@@ -782,10 +782,23 @@
     if (!chainparams.IsRegTestNet())
         throw std::runtime_error("mnconnect for regression testing (-regtest mode) only");
 
+    RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VARR});
+
+    Consensus::LLMQType llmq_type = Consensus::LLMQ_NONE;
+    if (request.params.size() > 2) {
+        RPCTypeCheckArgument(request.params[2], UniValue::VNUM);
+        llmq_type = (Consensus::LLMQType)request.params[2].get_int();
+    }
+
+    uint256 quorum_hash;
+    if (request.params.size() > 3) {
+        RPCTypeCheckArgument(request.params[3], UniValue::VSTR);
+        quorum_hash = uint256S(request.params[3].get_str());
+    }
+
     // First obtain the connection type
     const std::string& op_type = request.params[0].get_str();
     // Check provided mn_list
-    if (!request.params[1].isArray()) throw std::runtime_error("error: mnconnect arg1 must be an array of proreg txes hashes");
     const auto& array{request.params[1].get_array()};
     std::set<uint256> set_dmn_protxhash;
     for (unsigned int i = 0; i < array.size(); i++) {
@@ -801,13 +814,9 @@
         }
         return true;
     } else if (op_type == QUORUM_MEMBERS_CONN) {
-        Consensus::LLMQType llmq_type = (Consensus::LLMQType) request.params[2].get_int();
-        const uint256& quorum_hash = uint256S(request.params[3].get_str());
         mn_connan->setQuorumNodes(llmq_type, quorum_hash, set_dmn_protxhash);
         return true;
     } else if (op_type == IQR_MEMBERS_CONN) {
-        Consensus::LLMQType llmq_type = (Consensus::LLMQType) request.params[2].get_int();
-        const uint256& quorum_hash = uint256S(request.params[3].get_str());
         mn_connan->setMasternodeQuorumRelayMembers(llmq_type, quorum_hash, set_dmn_protxhash);
         return true;
     } else if (op_type == PROBE_CONN) {

```

File: src/init.cpp
Comment (furszy) [reply src/init.cpp:528]: left it in this way because down the commits line is extended with another flags in #2723.

**Comment by furszy** [comment]
Done, feedback tackled.

**Review by random-zebra — APPROVED** [review]
utACK ac014fba4ab5bb58e26d5db002e3ddf04fa9fe72

**Review by Fuzzbawls — APPROVED** [review]
ACK ac014fba4ab5bb58e26d5db002e3ddf04fa9fe72

**Pull request merged**
